File: Coinchange/coinchange.py
import sys
import time as t

LIMIT = 2**31 - 1
sys.setrecursionlimit(LIMIT)


MEMOIZATION = [-1] * 10000
VALUES = [1, 5, 6, 7]


def reset_memoization():
    global MEMOIZATION
    MEMOIZATION = [-1] * 10000


def coin_change_memo(n):
    global MEMOIZATION
    if n < 0:
        return 100000000
    elif n == 0:
        return 0
    elif MEMOIZATION[n] != -1:
        return MEMOIZATION[n]
    else:
        min_tot = min(coin_change_memo(n - VALUES[i]) + 1 for i in range(len(VALUES)))
        MEMOIZATION[n] = min(n, min_tot)
        return MEMOIZATION[n]

def coin_change_exhaustive(n):
    if n < 0:
        return 100000000
    elif n == 0:
        return 0
    else:
        min_temp = min(coin_change_exhaustive(n - VALUES[i]) + 1 for i in range(len(VALUES)))
        return min(n, min_temp)
    
def bottom_up(n):
    table = [0] * (n + 1)
    for i in range(1, n + 1):
        min_val = float('inf')
        for j in range(len(VALUES)):
            if i >= VALUES[j]:
                min_val = min(min_val, table[i - VALUES[j]] + 1)
        table[i] = min_val
    return table[n]

def test_limits_addition_memo():
    global VALUES
    VALUES = [1, 5, 6, 7]
    times = []
    n = 10
    delta = 0
    while delta < 1:  # 1 second
        t0 = t.time()
        number = coin_change_memo(n)
        delta = t.time() - t0
        reset_memoization()
        times.append((n, delta, number))
        n += 1
    return times


def test_limits_multiplication_memo():
    global VALUES
    VALUES = [1, 5, 6, 7]
    times = []
    n = 10
    delta = 0
    while delta < 1:  # 1 second
        t0 = t.time()
        number = coin_change_memo(n)
        delta = t.time() - t0
        reset_memoization()
        times.append((n, delta, number))
        n *= 2
    return times


def test_limits_addition(algo=coin_change_exhaustive, factor=1):
    global VALUES
    VALUES = [1, 5, 6, 7]
    times = []
    n = 10
    delta = 0
    while delta < 1:  # 1 second
        # print("add", n, delta)
        t0 = t.time()
        number = algo(n)
        delta = t.time() - t0
        times.append((n, delta, number))
        n += factor
    return times


def test_limits_multiplication(algo=coin_change_exhaustive):
    global VALUES
    VALUES = [1, 5, 6, 7]
    times = []
    n = 10
    delta = 0
    while delta < 1:  # 1 second
        # print("multi", n, delta)
        t0 = t.time()
        number = algo(n)
        delta = t.time() - t0
        times.append((n, delta, number))
        n *= 2
    return times


def plot_times(timesadd, timesmult, title):
    import matplotlib.pyplot as plt

    x = [t[0] for t in timesadd]
    y = [t[1] for t in timesadd]
    x_mult = [t[0] for t in timesmult]
    y_mult = [t[1] for t in timesmult]
    plt.title(title)
    plt.plot(x, y, color="red", marker="o", linestyle="solid")
    plt.plot(x_mult, y_mult, color="blue", marker="o", linestyle="solid")
    plt.legend(["Addition", "Multiplication"])
    plt.xlabel("n")
    plt.ylabel("time")
    plt.savefig(title + '.png')
    plt.show()





def ask_user_input():
    global VALUES
    n = int(input())
    a = int(input())
    b = int(input())
    c = int(input())
    VALUES = [1, a, b, c]
    return n


def main():
    n = ask_user_input()
    print(coin_change_memo(n))
    print(coin_change_exhaustive(n))
    print(bottom_up(n))


def test(algo):
    timesadd = test_limits_addition(algo, factor=10000)
    timesmult = test_limits_multiplication(algo)
    plot_times(timesadd, timesmult, "Coin Change Bottom Up")

test(bottom_up)

File: Coinchange/coinchange.cpp
#include <iostream>
#include <vector>
#include <limits>
#include <ctime>
#include <fstream>


struct coin_try {
    int amount;
    int result;
    double time;
};

float LIMIT = 1;
int VALUES[] = {1, 5, 6, 7};
std::vector<int> MEMOIZATION(1000, -1);

void reset_memoization(int n) {
    MEMOIZATION = std::vector<int>(n, -1);
}

int coin_change_memo(int n) {
    if (n < 0) {
        return 100000000;
    } else if (n == 0) {
        return 0;
    } else if (MEMOIZATION[n] != -1) {
        return MEMOIZATION[n];
    } else {
        int min_tot = std::numeric_limits<int>::max();
        for (int i = 0; i < sizeof(VALUES) / sizeof(VALUES[0]); ++i) {
            min_tot = std::min(min_tot, coin_change_memo(n - VALUES[i]) + 1);
        }
        MEMOIZATION[n] = std::min(n, min_tot);
        return MEMOIZATION[n];
    }
}


std::vector<coin_try> test_limits_addition() {
    std::vector<coin_try> times;
    coin_try coin;
    int n = 100;
    double delta = 0;
    while (delta < LIMIT/4) {
        reset_memoization(n);
        clock_t t0 = clock();
        int number = coin_change_memo(n);
        delta = (clock() - t0) / static_cast<double>(CLOCKS_PER_SEC);
        coin.amount = n;
        coin.result = number;
        coin.time = delta;
        times.push_back(coin);
        n += 100;
        if (n % 1000 == 0) {
            std::cout << n<< " | ";
            std::cout << delta << " s" << std::endl;
        }
    }
    return times;
}

std::vector<coin_try> test_limits_multiplication() {
    std::vector<coin_try> times;
    coin_try coin;
    int n = 5;
    double delta = 0;
    while (delta < LIMIT) {
        reset_memoization(n);
        clock_t t0 = clock();
        int number = coin_change_memo(n);
        delta = (clock() - t0) / static_cast<double>(CLOCKS_PER_SEC);
        coin.amount = n;
        coin.result = number;
        coin.time = delta;
        times.push_back(coin);
        n *= 2;
        if (n % 1000 < 50) {
            std::cout << n;
            std::cout << delta << " s" << std::endl;
        }
    }
    return times;
}



void print_in_file(std::vector<coin_try> times, std::string title){
    // Print the results in a file
    std::ofstream file;
    file.open(title);
    file << "amount,result,time\n";
    for (auto coin : times) {
        file << coin.amount << "," << coin.result << "," << coin.time << "\n";
    }
    file.close();
}

int main() {
    std::cout << "Starting tests" << std::endl;
    // int result = coin_change_memo(15);
    // std::cout << "Result: " << result << std::endl;

    auto times_memo_add = test_limits_addition();
    print_in_file(times_memo_add, "./outputs/memo_add.txt");
    std::cout << "Times_memo_Add done" << std::endl;

    auto times_memo_mult = test_limits_multiplication();
    print_in_file(times_memo_mult, "./outputs/memo_mult.txt");
    std::cout << "Times_memo_Mult done" << std::endl; 
    return 0;
}
